File: test/easysdk-release/ios/ReleaseSmoke/ReleaseSmokeApp.swift
import Foundation
import SwiftUI
import WuKongEasySDK

private enum SmokeError: Error {
    case missingEnvironment(String)
}

private final class MessageWaiter {
    private let lock = NSLock()
    private var continuation: CheckedContinuation<Message, Never>?
    private var pending: Message?

    func receive(_ message: Message) {
        lock.lock()
        if let continuation {
            self.continuation = nil
            lock.unlock()
            continuation.resume(returning: message)
            return
        }
        pending = message
        lock.unlock()
    }

    func wait() async -> Message {
        await withCheckedContinuation { continuation in
            lock.lock()
            if let pending {
                self.pending = nil
                lock.unlock()
                continuation.resume(returning: pending)
                return
            }
            self.continuation = continuation
            lock.unlock()
        }
    }
}

@MainActor
private final class ReleaseSmokeState: ObservableObject {
    @Published var status = "RUNNING"
    private var started = false

    func start() {
        guard !started else { return }
        started = true
        Task { await run() }
    }

    private func run() async {
        var sdk: WuKongEasySDK?
        do {
            let environment = ProcessInfo.processInfo.environment
            let aliceUid = try required("ALICE_UID", in: environment)
            let aliceToken = try required("ALICE_TOKEN", in: environment)
            let bobUid = try required("BOB_UID", in: environment)
            let aliceToBobText = try required("ALICE_TO_BOB_TEXT", in: environment)
            let bobToAliceText = try required("BOB_TO_ALICE_TEXT", in: environment)

            let client = try WuKongEasySDK.create(
                serverUrl: "ws://127.0.0.1:5200",
                uid: aliceUid,
                token: aliceToken,
                deviceFlag: .app
            )
            sdk = client
            let waiter = MessageWaiter()
            let listener = client.onMessage { message in
                guard
                    message.fromUid == bobUid,
                    message.payload["content"] as? String == bobToAliceText
                else { return }
                waiter.receive(message)
            }

            try await client.connect()
            let acknowledgment = try await client.send(
                channelId: bobUid,
                channelType: .person,
                payload: MessagePayload(type: 1, content: aliceToBobText)
            )
            let reply = await waiter.wait()

            client.removeListener(listener)
            client.disconnect()

            guard acknowledgment.messageSeq > 0, reply.messageSeq > 0, !client.isConnected else {
                throw NSError(domain: "ReleaseSmoke", code: 1)
            }

            try writeReceipt([
                "status": "PASS",
                "package": "WuKongEasySDK CocoaPods 1.1.1",
                "platform": "iOS Simulator",
                "aliceToBob": true,
                "bobToAlice": true,
                "disconnected": true,
            ])
            status = "PASS"
            print("IOS_RELEASE_SMOKE_PASS package=WuKongEasySDK@1.1.1")
        } catch {
            sdk?.disconnect()
            try? writeReceipt([
                "status": "FAIL",
                "errorType": String(describing: type(of: error)),
            ])
            status = "FAIL"
            print("IOS_RELEASE_SMOKE_FAIL error=\(type(of: error))")
        }
    }

    private func required(_ name: String, in environment: [String: String]) throws -> String {
        guard let value = environment[name], !value.isEmpty else {
            throw SmokeError.missingEnvironment(name)
        }
        return value
    }

    private func writeReceipt(_ receipt: [String: Any]) throws {
        let directory = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask)[0]
        let data = try JSONSerialization.data(withJSONObject: receipt, options: [.sortedKeys])
        try data.write(to: directory.appendingPathComponent("release-smoke.json"), options: .atomic)
    }
}

private struct ContentView: View {
    @StateObject private var state = ReleaseSmokeState()

    var body: some View {
        Text(state.status)
            .task { state.start() }
    }
}

@main
struct ReleaseSmokeApp: App {
    var body: some Scene {
        WindowGroup { ContentView() }
    }
}

File: test/easysdk-release/flutter/release_smoke_app.dart
import 'dart:async';
import 'dart:convert';
import 'dart:io';

import 'package:flutter/material.dart';
import 'package:wukong_easy_sdk/wukong_easy_sdk.dart';

void main() {
  WidgetsFlutterBinding.ensureInitialized();
  runApp(const ReleaseSmokeApp());
  unawaited(_runReleaseSmoke());
}

class ReleaseSmokeApp extends StatelessWidget {
  const ReleaseSmokeApp({super.key});

  @override
  Widget build(BuildContext context) {
    return const MaterialApp(
      home: Scaffold(body: Center(child: Text('WuKongEasySDK release smoke'))),
    );
  }
}

String _requiredConfig(Map<String, dynamic> config, String name) {
  final value = config[name];
  if (value is! String || value.isEmpty) {
    throw StateError('Missing required release smoke setting: $name');
  }
  return value;
}

Future<Map<String, dynamic>> _readConfig(File configFile) async {
  final source = await configFile.readAsString();
  await configFile.delete();
  final decoded = jsonDecode(source);
  if (decoded is! Map<String, dynamic>) {
    throw const FormatException('Release smoke config must be a JSON object');
  }
  return decoded;
}

Future<void> _writeReceipt(File receipt, Map<String, Object> payload) async {
  final temporary = File('${receipt.path}.tmp');
  await temporary.writeAsString(jsonEncode(payload), flush: true);
  await temporary.rename(receipt.path);
}

Future<void> _runReleaseSmoke() async {
  final configFile = File(
    '${Directory.systemTemp.path}/release-smoke-config.json',
  );
  final receipt = File('${Directory.systemTemp.path}/release-smoke.json');
  final sdk = WuKongEasySDK.getInstance();
  WuKongEventListener<Message>? listener;
  var stage = 'config';

  try {
    final config = await _readConfig(configFile);
    final aliceUid = _requiredConfig(config, 'ALICE_UID');
    final aliceToken = _requiredConfig(config, 'ALICE_TOKEN');
    final bobUid = _requiredConfig(config, 'BOB_UID');
    final aliceToBobText = _requiredConfig(config, 'ALICE_TO_BOB_TEXT');
    final bobToAliceText = _requiredConfig(config, 'BOB_TO_ALICE_TEXT');
    final receivedReply = Completer<Message>();

    listener = (message) {
      final payload = message.payload;
      if (message.fromUid == bobUid &&
          payload is Map &&
          payload['content'] == bobToAliceText &&
          !receivedReply.isCompleted) {
        receivedReply.complete(message);
      }
    };
    sdk.addEventListener(WuKongEvent.message, listener);

    stage = 'init';
    await sdk
        .init(
          WuKongConfig(
            serverUrl: 'ws://127.0.0.1:5200',
            uid: aliceUid,
            token: aliceToken,
            deviceFlag: WuKongDeviceFlag.app,
          ),
        )
        .timeout(const Duration(seconds: 30));
    stage = 'connect';
    await sdk.connect().timeout(const Duration(seconds: 30));
    if (!sdk.isConnected) {
      throw StateError('SDK did not report a connected state');
    }

    stage = 'send';
    final acknowledgment = await sdk
        .send(
          channelId: bobUid,
          channelType: WuKongChannelType.person,
          payload: {'type': 1, 'content': aliceToBobText},
        )
        .timeout(const Duration(seconds: 30));
    if (acknowledgment.messageSeq <= 0) {
      throw StateError(
        'Send acknowledgment did not contain a message sequence',
      );
    }

    stage = 'reply';
    final reply = await receivedReply.future.timeout(
      const Duration(seconds: 30),
    );
    if (reply.messageSeq <= 0) {
      throw StateError('Reply did not contain a message sequence');
    }

    stage = 'disconnect';
    sdk.disconnect();
    await Future<void>.delayed(const Duration(milliseconds: 200));
    if (sdk.isConnected) {
      throw StateError('SDK remained connected after disconnect');
    }

    await _writeReceipt(receipt, {
      'status': 'PASS',
      'package': 'wukong_easy_sdk@1.1.0',
      'alice_to_bob': true,
      'bob_to_alice': true,
      'disconnected': true,
    });
    // ignore: avoid_print
    print(
      'FLUTTER_RELEASE_SMOKE_PASS package=wukong_easy_sdk@1.1.0 '
      'alice-to-bob=true bob-to-alice=true disconnected=true',
    );
  } catch (error) {
    await _writeReceipt(receipt, {
      'status': 'FAIL',
      'stage': stage,
      'error_type': error.runtimeType.toString(),
    });
    // ignore: avoid_print
    print('FLUTTER_RELEASE_SMOKE_FAIL type=${error.runtimeType}');
  } finally {
    if (listener != null) {
      sdk.removeEventListener(WuKongEvent.message, listener);
    }
    if (sdk.isConnected) {
      sdk.disconnect();
    }
  }
}
